### PR TITLE
fix(cloudfront): bucket policy for Origin Access Identities is overly permissive

### LINKED ISSUE #13086: (cloudfront): default origin access identity policies grant unneeded permissions
When creating a CloudFront Distribution with a S3Origin, a bucket policy granting the following actions is attached:

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": "arn:aws:iam::cloudfront:user/CloudFront Origin Access Identity xyz"
            },
            "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*"
            ],
            "Resource": [
                "arn:aws:s3::: xyz",
                "arn:aws:s3:::xyz/*"
            ]
        }
    ]
}
```

This however is far more than the advised [s3:GetObject](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-restricting-access-to-s3.html#private-content-granting-permissions-to-oai) action. Specifically this will lead to a xml listing of all bucket objects when accessing the CloudFront domain's root object, _if no default root object is set_.

The same policy is applied whether an `S3Origin` is used with `Distribution`, or if a bucket origin and origin access identity are used with `CloudFrontWebDistribution`.

### PR DESCRIPTION
**Pull request by njlynch**
For both `CloudFrontWebDistribution` and `Distribution`, if an origin access
identity (OAI) is used to access a S3 bucket origin, the default
`bucket.grantRead()` method is used to grant read access to the bucket for the
OAI. `grantRead` will create a statement in the bucket policy that grants:
`s3:GetObject*`, `s3:GetBucket*`, and `s3:List*`
to the OAI.

The [official documentation](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-restricting-access-to-s3.html) in the developer guide
states that to allow the OAI to read objects in the bucket, only `s3:GetObject`
is necessary.

The impact of the additional permissions is that if a user navigates to the root
of the distribution (e.g., https://d12345abcdef.cloudfront.net/) *AND* no default
root object (e.g., index.html) is set for the distribution, the user will receive
an XML list of the objects in the bucket. This is likely undesired behavior.

This same policy has been the behavior for OAI + S3 buckets since the first
introduction of the `CloudFrontWebDistribution` construct years ago. However,
the `CloudFrontWebDistribution` also always defaults the root object to
'index.html' (see #3486 for discussion on that). The side-effect of this is
that unless the user selects to use an OAI and overrides the default root object
to an empty string, visiting the root will result either in valid content or an
AccessDenied message.

However, since the `Distribution` constructs do *not* default the root object,
exposing the bucket index becomes much more likely.

This change restricts the permissions granted to the OAI user to just the
necessary `s3:GetObject` permissions.

fixes #13086

---

**NOTE:** This PR corrects the behavior without introducing a new feature flag.
Any customers who rely on the bucket listing behavior would be broken by this
change. However, I anticipate that represents a very small minority, versus the
number of customers who may accidentally expose their bucket contents. I would
very much like to have a discussion about whether this is the correct choice.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by skinny85 — APPROVED** [review]
LGTM! I agree with you that I don't think we need a feature flag for this.

**Review by nija-at — COMMENTED** [review]
Feature flag is definitely the safest option here. If we feel that customers are unlikely to depend on this behaviour, it might be prudent to skip this.

Could we at least provide an option on the `Distribution` construct for users to restore the old behaviour?

File: packages/@aws-cdk/aws-cloudfront-origins/lib/s3-origin.ts [hunk 1]
```diff
@@ -79,7 +80,16 @@ class S3BucketOrigin extends cloudfront.OriginBase {
       this.originAccessIdentity = new cloudfront.OriginAccessIdentity(oaiScope, oaiId, {
         comment: `Identity for ${options.originId}`,
       });
-      this.bucket.grantRead(this.originAccessIdentity);
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cloudfront-origins/lib/s3-origin.ts:82]: Isn't there a grant() API on bucket that takes a list of actions?

Comment (njlynch) [reply packages/@aws-cdk/aws-cloudfront-origins/lib/s3-origin.ts:82]: There's a `private` one:
https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/aws-s3/lib/bucket.ts#L716

I briefly toyed with the idea of adding a `grantReadObject` method, but felt it was a bit niche.

**Comment by njlynch** [comment]
> Could we at least provide an option on the Distribution construct for users to restore the old behaviour?

I'd rather not, unless we get significant pushback. Exposing the bucket contents at the distribution root doesn't seem like a typically-desirable behavior, and there's a reasonable workaround.

```ts
// Before - Does the bucket listing by default
const bucket = new s3.Bucket(this, 'Bucket');
const s3Origin = new origins.S3Origin(this, 'S3Origin', { bucket });

// After - Customers who want to intentionally have the bucket listing
const bucket = new s3.Bucket(this, 'Bucket');
const originAccessIdentity = new cloudfront.OriginAccessIdentity(this, 'OAI');
bucket.grantRead(originAccessIdentity);
const s3Origin = new origins.S3Origin(this, 'S3Origin', { bucket, originAccessIdentity });
```

File: packages/@aws-cdk/aws-cloudfront-origins/lib/s3-origin.ts (hunk 1 above)
Comment (nija-at) [reply packages/@aws-cdk/aws-cloudfront-origins/lib/s3-origin.ts:82]: It's a common pattern for constructs to expose a `grant(grantee, ...actions)` API for such niche use cases.

**Pull request merged**